src/main/java/org/truffleruby/core/numeric/IntegerNodes.java

```diff
@@ -1755,6 +1755,17 @@ protected int getLimit() {
 
     }
 
+    @Primitive(name = "mod_pow", lowerFixnum = { 0, 1, 2 })
+    public abstract static class ModPowNode extends PrimitiveArrayArgumentsNode {
+        @Child private FixnumOrBignumNode fixnumOrBignum = new FixnumOrBignumNode();
+
+        @Specialization
```

Comment:
I am not that familiar with the Truffle framework, so I'm just guessing that this single specialization is enough. However, from the Ruby world `pow` is implemented on `Integer`, the other two arguments also have to be `Integer` to not get an exception before calling this primitive.

Reply:
`Integer` here means the Java boxed `int`, not Ruby `Integer`s.
We need specialization for both `long` and `DynamicObject + isRubyBignum(argument)` since both are possible.
This is unfortunate as it results in quite a few combinations in this case.
A way to deal with that and have a single specialization would be to use `@CreateCast` on each argument, and add a new `BigIntegerCastNode`, similar to `LongCastNode`, that would accept `int`, `long` and Bignums and return a BigInteger for it.